[Durable Objects] Add documentation for gradual deployments for Durable Objects.

jp4a50 opened the pull request.

jp4a50 commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
````diff
@@ -91,6 +91,53 @@ done
 ```
 You should see 10 responses. Responses will reflect the content returned by the versions in your deployment. Responses will vary depending on the percentages configured in step #6. 
 
+## Gradual deployments for Durable Objects
````
@taylorlee we need to document any DO-specific limitations.

One is that you cannot combine a gradual deployment and a DO migration, right?

Anything else?

jp4a50 replied:
@taylorlee ping

tanushreesharma-cf replied:
Added this limitation to the limits section below

jp4a50 commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,53 @@ done
 
+## Gradual deployments for Durable Objects
+
+Due to [global uniqueness](durable-objects/platform/known-issues/#global-uniqueness), only one version of each [Durable Object](/durable-objects/) can run at a time. This means that gradual deployments work slightly differently for Durable Objects.
```
@tanushree-sharma you added this to the version affinity section before I closed that PR:

> By default, the percentages configured when using gradual deployments operate on a per-request basis — a request has a X% probability of invoking one of two versions of the Worker in the [deployment](/workers/configuration/versions-and-deployments/#deployments). 

I think this deserves to go into the general gradual deployments documentation (in a section preceeding the Durable Objects one), then my claim that "gradual deployments work slightly differently for Durable Objects" will have a point of reference.

vy-ton commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,53 @@ done
+- For a given deployment, requests to each Durable Object will always use the same Worker version.
+- When you specify each version in the same order as the previous deployment and increase the percentage of a version, Durable Objects which were previously assigned that version will not be assigned a different version. In our example, Durable Object "foo" would never revert from version "B" to version "A".
+- The Durable Object will only be [reset](/durable-objects/reference/troubleshooting/#durable-object-reset-because-its-code-was-updated) when it is assigned a different version, so each Durable Object will only be reset once in our example.
+
```
Should we add a bullet for Websockets not being disconnected with gradual rollouts? I believe we can't remove https://github.com/cloudflare/cloudflare-docs/blob/production/content/durable-objects/reference/websockets.md?plain=1#L23 just yet

jp4a50 replied:
Websockets will still be disconnected whenever a specific Durable Object's code is updated. The difference is that a gradual deployment may only update the code for a subset of DOs now, so only websockets to those DOs will be disconnected.

I think ideally we should have a dedicated section on Durable Object resets which defines when they happen and their side effects (e.g. websockets being disconnected) and both the above documentation and the documentation in your link should reference that.

What d'you think?

vy-ton replied:
Makes sense, I'll track the DO reset behavior docs.

JoshuaClaeys commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,53 @@ done
+| bar            | A       |
+| baz            | A       |
+
+Now you create a gradual deployment, setting version "A" to 20% and version "B" to 80%.
```
```suggestion
Now you create a gradual deployment, setting version "A" to 80% and version "B" to 20%.
```

JoshuaClaeys commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
````diff
@@ -91,6 +91,53 @@ done
 ```
 You should see 10 responses. Responses will reflect the content returned by the versions in your deployment. Responses will vary depending on the percentages configured in step #6. 
 
+## Gradual deployments for Durable Objects
````
I would love see a section that directly addresses forwards and backwards compatibility requirements between eyeball Worker and Durable Objects during a release (old eyeball talks to new DO and new eyeball talks to old DO).

If this is already mentioned elsewhere outside of gradual deployments page (if not, I think it should be), this could be as simple as a link and mentioning that this easily missed compatibility issue will be exacerbated with gradual deployments.

jp4a50 replied:
Good call. The requirement for forwards and backwards compatibility for durable objects and the worker that talks to them is briefly mentioned [here](https://developers.cloudflare.com/durable-objects/platform/known-issues/#code-updates) so I will link to that for now and elaborate on how gradual deployments affects this problem.

I think we could generally do with a section explaining compatibility requirements more generally. I started drafting a new section in the "Versions & Deployments" page that was added in @tanushree-sharma 's [PR](https://github.com/cloudflare/cloudflare-docs/pulls/tanushreesharma-cf) but it's tricky to explain this in general terms without introducing new terminology and/or using diagrams/example code so I'll leave that for another time.

JoshuaClaeys commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,53 @@ done
+
+Due to [global uniqueness](durable-objects/platform/known-issues/#global-uniqueness), only one version of each [Durable Object](/durable-objects/) can run at a time. This means that gradual deployments work slightly differently for Durable Objects.
+
+When you create a new gradual deployment for a Worker that defines a Durable Object, each Durable Object is assigned a Worker version based on the percentages you configured in your [deployment](/workers/configuration/deployments/). This version will not change until you create a new deployment.
```
The wording of "When you create a new gradual deployment for a Worker that **defines a Durable Object**, ..." threw me off for a second. I read it as "a deployment that defines a Durable Object", i.e. deploying a change that includes a newly defined Durable Object, which isn't supported.

jp4a50 replied:
Yeah, I see how it's ambiguous. Reworded.

irvinebroque reviewed (approved):
Merge conflicts then 🚀

irvinebroque commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,61 @@ done
+| bar            | A       |
+| baz            | A       |
+
+Now you create a gradual deployment, setting version "A" to 80% and version "B" to 20%.
```
Language — I'm not sure we want to say "create a gradual deployment" but rather, "when you deploy two versions of your Durable Object at once, setting Version "A to 80% and Version "B" to 20%"

jp4a50 replied:
I was trying to be consistent with the language in [this section](https://developers.cloudflare.com/workers/configuration/versions-and-deployments/gradual-deployments/#3-create-a-new-deployment), but I suppose you could argue that the "create a deployment" phrasing is warranted in that "how to"-style section but not this one, so will change it.

irvinebroque commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -91,6 +91,61 @@ done
+
+Typically, your Durable Object Worker will define both your Durable Object class and the Worker that interacts with it. In this case, you cannot deploy changes to your Durable Object and its Worker independently.
+
+[You are advised](/durable-objects/platform/known-issues/#code-updates) to ensure that API changes between your Durable Object and its Worker are forwards and backwards compatible whether you are using gradual deployments or not. However, using gradual deployments will make it even more likely that different versions of your Durable Objects and its Worker will interact with each other.
```
```suggestion
[You should](/durable-objects/platform/known-issues/#code-updates) ensure that API changes between your Durable Object and its Worker are forwards and backwards compatible, regardless of whether you are using gradual deployments or not. However, using gradual deployments will make it even more likely that different versions of your Durable Objects and its Worker will interact with each other.
```

ndisidore commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -0,0 +1,190 @@
+
+{{<Aside type="note">}}
+
+Gradual deployments are in **beta and under active development**. Please read [limitations](/workers/configuration/versions-and-deployments//gradual-deployments/#limitations) before using this feature.
```
nit: 2 `/` in this url

deadlypants1973 commented:
This PR seems to be carrying over content that was already edited, reviewed and pushed last week on https://github.com/cloudflare/cloudflare-docs/pull/13429 and rewriting it to its pre-reviewed state.

Does this need a rebase? @jp4a50

jp4a50 commented:
> This PR seems to be carrying over content that was already edited, reviewed and pushed last week on #13429 and rewriting it to its pre-reviewed state.
> 
> Does this need a rebase? @jp4a50

Yep, it was because I originally targeted the branch used for #13429 which was squashed, merged and deleted. I've rebased now.

irvinebroque commented on content/durable-objects/platform/known-issues.md:
```diff
@@ -19,7 +19,7 @@ A Durable Object may be replaced in the event of a network partition or a softwa
 ## Code updates
 
-Code changes for Workers and Durable Objects are released globally in an eventually consistent manner. Because each Durable Object is globally unique, the situation can arise that a request arrives to the latest version of your Worker (running in one part of the world), which then calls to a unique Durable Object running the previous version of your code for a short period of time (typically seco ...
+Code changes for Workers and Durable Objects are released globally in an eventually consistent manner. Because each Durable Object is globally unique, the situation can arise that a request arrives to the latest version of your Worker (running in one part of the world), which then calls to a unique Durable Object running the previous version of your code for a short period of time. If you create  ...
```
Probably keep the "typiccally seconds to minutes" piece for the base case scenario

Maddy-Cloudflare commented on content/durable-objects/configuration/durable-object-gradual-deployments.md:
```diff
@@ -0,0 +1,13 @@
+---
+pcx_content_type: navigation
+title: Gradual Deployments
```
```suggestion
title: Gradual deployments
```

jp4a50 replied:
We usually capitalize Deployments elsewhere so leaving as-is.

Maddy-Cloudflare commented on content/durable-objects/platform/known-issues.md:
```diff
@@ -19,7 +19,7 @@ A Durable Object may be replaced in the event of a network partition or a softwa
 ## Code updates
 
-Code changes for Workers and Durable Objects are released globally in an eventually consistent manner. Because each Durable Object is globally unique, the situation can arise that a request arrives to the latest version of your Worker (running in one part of the world), which then calls to a unique Durable Object running the previous version of your code for a short period of time (typically seco ...
+Code changes for Workers and Durable Objects are released globally in an eventually consistent manner. Because each Durable Object is globally unique, the situation can arise that a request arrives to the latest version of your Worker (running in one part of the world), which then calls to a unique Durable Object running the previous version of your code for a short period of time (typically seco ...
```
```suggestion
Code changes for Workers and Durable Objects are released globally in an eventually consistent manner. Because each Durable Object is globally unique, the situation can arise that a request arrives to the latest version of your Worker (running in one part of the world), which then calls to a unique Durable Object running the previous version of your code for a short period of time (typically seconds to minutes). If you create a [gradual deployment](/workers/configuration/versions-and-deployments/gradual-deployments/), this period of time is determined by how long your live deployment is configured to use more than one version.
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+
+### Example
+
+Imagine that you have previously created 3 Durable Objects, [deriving their IDs from the names](/durable-objects/configuration/access-durable-object-from-a-worker/#derive-ids-from-names) "foo", "bar" and "baz".
```
```suggestion
This example assumes that you have previously created 3 Durable Objects and [derived their IDs from the names](/durable-objects/configuration/access-durable-object-from-a-worker/#derive-ids-from-names) "foo", "bar" and "baz".
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+
+Imagine that you have previously created 3 Durable Objects, [deriving their IDs from the names](/durable-objects/configuration/access-durable-object-from-a-worker/#derive-ids-from-names) "foo", "bar" and "baz".
+
+Your Worker is currently on a version that we will call version "A" and you want to gradually deploy a new version "B" of your Worker.
```
```suggestion
Your Worker is currently on a version called version "A" and you want to gradually deploy a new version "B" of your Worker.
```

jp4a50 replied:
I'd like to leave this phrasing as-is to convey that this is just a made-up identifier, rather than the actual version ID.

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+
+Your Worker is currently on a version that we will call version "A" and you want to gradually deploy a new version "B" of your Worker.
+
+Here's how the versions of your Durable Objects might change as you progress your gradual deployment:
```
```suggestion
Here is how the versions of your Durable Objects might change as you progress your gradual deployment:
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+| Version B: 50% <br> Version A: 50%     | B     | B     | A     |
+| Version B: 100% <br>                   | B     | B     | B     |
+
+This is only an example, so the versions assigned to your Durable Objects may be different. However, the following is guaranteed:
```
```suggestion
This is only an example, so the versions assigned to your Durable Objects may be different. However, the following is guaranteed:

```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+
+This is only an example, so the versions assigned to your Durable Objects may be different. However, the following is guaranteed:
+- For a given deployment, requests to each Durable Object will always use the same Worker version.
+- When you specify each version in the same order as the previous deployment and increase the percentage of a version, Durable Objects which were previously assigned that version will not be assigned a different version. In our example, Durable Object "foo" would never revert from version "B" to version "A".
```
```suggestion
- When you specify each version in the same order as the previous deployment and increase the percentage of a version, Durable Objects which were previously assigned that version will not be assigned a different version. In this example, Durable Object "foo" would never revert from version "B" to version "A".
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+This is only an example, so the versions assigned to your Durable Objects may be different. However, the following is guaranteed:
+- For a given deployment, requests to each Durable Object will always use the same Worker version.
+- When you specify each version in the same order as the previous deployment and increase the percentage of a version, Durable Objects which were previously assigned that version will not be assigned a different version. In our example, Durable Object "foo" would never revert from version "B" to version "A".
+- The Durable Object will only be [reset](/durable-objects/reference/troubleshooting/#durable-object-reset-because-its-code-was-updated) when it is assigned a different version, so each Durable Object will only be reset once in our example.
```
```suggestion
- The Durable Object will only be [reset](/durable-objects/reference/troubleshooting/#durable-object-reset-because-its-code-was-updated) when it is assigned a different version, so each Durable Object will only be reset once in this example.
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -148,6 +148,43 @@ Selected operation under **Modify request header**: _Set dynamic_
+
+Typically, your Durable Object Worker will define both your Durable Object class and the Worker that interacts with it. In this case, you cannot deploy changes to your Durable Object and its Worker independently.
+
+[You should](/durable-objects/platform/known-issues/#code-updates) ensure that API changes between your Durable Object and its Worker are forwards and backwards compatible whether you are using gradual deployments or not. However, using gradual deployments will make it even more likely that different versions of your Durable Objects and its Worker will interact with each other.
```
```suggestion
You should ensure that API changes between your Durable Object and its Worker are [forwards and backwards compatible](/durable-objects/platform/known-issues/#code-updates) whether you are using gradual deployments or not. However, using gradual deployments will make it even more likely that different versions of your Durable Objects and its Worker will interact with each other.
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -194,5 +231,6 @@ You can only create a new deployment with the last 10 uploaded versions of your
-- Gradual deployments is not supported for Workers with the [mTLS binding](/workers/runtime-apis/bindings/mtls/).
-- Creating a gradual deployment with two different configurations for [Smart Placement](/workers/configuration/smart-placement/) is not supported.
\ No newline at end of file
+- Gradual deployments is not supported for Workers with the [mTLS binding](/workers/runtime-apis/bindings/mtls/). You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with an mTLS binding. 
```
```suggestion
- Gradual deployments are not supported for Workers with the [mTLS binding](/workers/runtime-apis/bindings/mtls/). Use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with an mTLS binding. 
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -194,5 +231,6 @@ You can only create a new deployment with the last 10 uploaded versions of your
-- Creating a gradual deployment with two different configurations for [Smart Placement](/workers/configuration/smart-placement/) is not supported.
\ No newline at end of file
+- Gradual deployments is not supported for Workers with the [mTLS binding](/workers/runtime-apis/bindings/mtls/). You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with an mTLS binding. 
+- Creating a gradual deployment with two different configurations for [Smart Placement](/workers/configuration/smart-placement/) is not supported. You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with Smart Placement enabled. 
```
```suggestion
- Creating a gradual deployment with two different configurations for [Smart Placement](/workers/configuration/smart-placement/) is not supported. Use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with Smart Placement enabled. 
```

Maddy-Cloudflare commented on content/workers/configuration/versions-and-deployments/gradual-deployments.md:
```diff
@@ -194,5 +231,6 @@ You can only create a new deployment with the last 10 uploaded versions of your
\ No newline at end of file
+- Gradual deployments is not supported for Workers with the [mTLS binding](/workers/runtime-apis/bindings/mtls/). You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with an mTLS binding. 
+- Creating a gradual deployment with two different configurations for [Smart Placement](/workers/configuration/smart-placement/) is not supported. You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) for Workers with Smart Placement enabled. 
+- Creating a gradual deployment with [Durable Object migrations](/durable-objects/reference/durable-objects-migrations/) is not supported. You must use [`wrangler deploy`](/workers/wrangler/commands/#deploy) if you are applying a DO migration. 
```
```suggestion
- Creating a gradual deployment with [Durable Object migrations](/durable-objects/reference/durable-objects-migrations/) is not supported. Use [`wrangler deploy`](/workers/wrangler/commands/#deploy) if you are applying a [Durable Object migration](/durable-objects/reference/durable-objects-migrations/). 
```

irvinebroque commented on content/durable-objects/configuration/durable-object-gradual-deployments.md:
```diff
@@ -0,0 +1,13 @@
+title: Gradual Deployments
+weight: 25
+
+external_link: /workers/configuration/versions-and-deployments/gradual-deployments/#gradual-deployments-for-durable-objects
```
adding this in DO was nice touch
